### ISSUE TITLE
Creating and loading custom pipeline components example

### ISSUE THREAD

**Issue by nyejon** [explosion/spaCy#2947]
Hi, I have been struggling to deal with the loading of a custom pipeline component after saving the model. I found this solution here: https://github.com/explosion/spaCy/issues/2682#issuecomment-420628360

Would it be a good idea to update the docs to suggest users do something like this to prevent the issue of requiring the "nlp" object to initialize the pipe when loading from disk? (Kind of a chicken and egg situation) Perhaps I am overlooking something and there is already a better solution.

Therefore this:

```
class EntityMatcher(object):
    name = 'entity_matcher'

    def __init__(self, nlp, terms, label):
        patterns = [nlp(text) for text in terms]
        self.matcher = PhraseMatcher(nlp.vocab)
        self.matcher.add(label, None, *patterns)

    def __call__(self, doc):
        matches = self.matcher(doc)
        for match_id, start, end in matches:
            span = Span(doc, start, end, label=match_id)
            doc.ents = list(doc.ents) + [span]
        return doc
```

```
class EntityMatcher(object):
    name = 'entity_matcher'

    def __init__(self, nlp, terms, label):
        self.nlp = nlp
        self.terms = terms
        self.label = label
        self.initialized = False
        self.matcher = PhraseMatcher(self.nlp.vocab)

    def __call__(self, doc):
        # Initialize the pipeline component on the first call
        if not self.initialized:
            self.initialized = True
            patterns = [self.nlp(text) for text in self.labelterms]
            self.matcher.add(label, None, *patterns)
            
        matches = self.matcher(doc)
        for match_id, start, end in matches:
            span = Span(doc, start, end, label=match_id)
            doc.ents = list(doc.ents) + [span]
        return doc
```

The source can be found here:
https://spacy.io/usage/processing-pipelines#custom-components

Edit: The self.initialized = True needs to be set directly after the if statement before nlp is called.

**Comment by ines** [comment]
Yes, that's definitely an option!

Alternatively, you could also add an entry to the `Language.factories`, which is a writable dictionary where spaCy looks up how to initialize pipeline components. For example:

```python
from spacy.language import Language

Language.factories['entity_matcher'] = lambda nlp, **cfg: EntityMatcher(nlp)
```

The factory receives the shared `nlp` object, as well as the config keyword arguments that are passed to `spacy.load`. So you could even do something like this to make the `entity_matcher_label` available as a keyword argument in your `EntityMatcher` factory:

```python
nlp = spacy.load('some_custom_model', entity_matcher_label='SOME_LABEL')
```

For a more detailed end-to-end example of how to package custom models or components in a model, you might find [my comment on this thread](https://github.com/explosion/spaCy/issues/2676#issuecomment-413274806) helpful. We're also hoping that we can make this process easier in the upcoming version(s).

Btw, another related thing to check out: If you're working with rule-based NER (or a combination of rule-based and statistical NER), you can also try my new `EntityRuler` component in the currently v2.1.x nightly version. See #2513 for details and documentation. The nightly build is available as `spacy-nightly`, or on the `develop` branch if you want to build the latest state from source.

**Comment by nyejon** [comment]
Hi Ines,

Thanks, I had already added the entity_matcher to the Language.factories and that is where it required the nlp object just to initialise the entity matcher and run the terms through the nlp object. 

In the #2682 you stated that the weights are not loaded yet and that only after everything is initialised the weights would be loaded. If I just added the entity_matcher to the Language.factories then I would get a "bool has no tok2vec" error or something similar. This is not a problem when using the nlp.add_pipe() method as the nlp object already has the weights loaded.

I will definitely look into 2.1 and the EntityRuler - thanks for the tip!

**Comment by ines** [comment]
Ahh, sorry, I think I missed that. (That `bool has no xxx` stuff will also be fixed in 2.1 btw.) 

I wonder what happens if you just use `nlp.make_doc(text)` to create the patterns? In fact, this should probably be fixed in the docs as well, since the `PhraseMatcher` really only needs a tokenized `Doc` object, not a fully processed one. Tagging and parsing it is unnecessary and just takes longer.

**Comment by ines** [comment]
Yes, the nightly should be good in development – in fact, it's always super helpful to have more people test it in "real life" so we can make sure there are no bugs or regressions.

The latest alpha version is `v2.1.0a2` ([see diff](https://github.com/explosion/spaCy/compare/v2.1.0a1...v2.1.0a2)) and mostly includes internals, e.g. allowing pre-training (#2931). I'm not 100% sure if the alpha models currently work with it or if we have to train new ones, so `v2.1.0a1` should be safest for now.
